pkg/api/norman/customization/cluster/validator.go

```diff
@@ -440,25 +443,73 @@ func validateEKSAccess(request *types.APIContext, eksConfig map[string]interface
 	return nil
 }
 
-func validateEKSNodegroups(request *types.APIContext, eksConfig map[string]interface{}, prevSpec *eksv1.EKSClusterConfigSpec) error {
-	noNodegroupsError := httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("must have at least one nodegroup"))
-	nodegroups, ok := eksConfig["nodeGroups"].([]interface{})
-	if !ok || nodegroups == nil {
-		if request.Method == http.MethodPost {
-			return noNodegroupsError
+// validateEKSKubernetesVersion checks whether a kubernetes version is provided and if it is supported
+func validateEKSKubernetesVersion(spec *v32.ClusterSpec, prevCluster *v3.Cluster) error {
+	clusterVersion := spec.EKSConfig.KubernetesVersion
+	if clusterVersion == nil {
+		return nil
+	}
+
+	if aws.StringValue(clusterVersion) == "" {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, "cluster kubernetes version cannot be empty string")
+	}
+
+	if prevCluster == nil {
+		return nil
+	}
+
+	nodegroups := prevCluster.Spec.EKSConfig.NodeGroups
+	if nodegroups == nil {
+		if prevCluster.Status.EKSStatus.UpstreamSpec == nil {
+			return httperror.NewAPIError(httperror.ClusterUnavailable, "upstream spec not available yet, unable to validate kubernetes version upgrade")
 		}
-		if prevSpec.NodeGroups != nil && len(prevSpec.NodeGroups) == 0 {
-			return noNodegroupsError
+		nodegroups = prevCluster.Status.EKSStatus.UpstreamSpec.NodeGroups
+	}
+
+	currentVersion := prevCluster.Spec.EKSConfig.KubernetesVersion
+	if currentVersion == nil {
+		currentVersion = prevCluster.Status.EKSStatus.UpstreamSpec.KubernetesVersion
+	}
+	for _, ng := range nodegroups {
+		if aws.StringValue(currentVersion) != aws.StringValue(ng.Version) {
+			return httperror.NewAPIError(httperror.InvalidBodyContent, "nodegroups' versions must match cluster's kubernetes version before cluster upgrade")
```

Comment:
How come the `clusterVersion` from spec is never checked here ? All we do is make sure its not blank

Reply:
All we are doing here is validating that the nodegroup version and current cluster version are equal before allowing an attempted upgrade. It being in the request tells us an upgrade is being attempted. If the version is unsupported or something, a controller error will appear.

Reply:
I was confused about purpose here, we are really checking the old versions to make sure they line up first, and if those line up we can proceed. Might be nice to comment that inline somewhere.